syntax = "proto3";

package capkfa;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

option go_package = "github.com/huuhait/cheetah/pkg/pb/capkfa;pb";

message CreateSessionRequest {
  string key = 1;
  string hwid = 2;
  string version = 3;
}

message CreateSessionResponse {
  string session_id = 1;
  bool valid = 2;
  int64 expired_at = 3; // Duration in seconds
}

message GetSessionRequest {
  string key = 1;
}

message GetSessionResponse {
  string key = 1;
  string session_id = 2;
  google.protobuf.Timestamp created_at = 3;
  google.protobuf.Timestamp last_ping = 4;
}

message KillAllSessionsRequest {
  string key = 1;
}

message KillAllSessionsResponse {
  repeated string killed_session_ids = 1;
}

message PingRequest {
  string key = 1;
  string hwid = 2;
  string version = 3;
  string session_id = 4;
}

message PingResponse {
  string session_id = 1;
  bool valid = 2;
}

message GetStatusResponse {
  bool online = 1;
  string version = 2;
}

message RemoteConfigMouseServer {
  string uri = 1;
}

message RemoteConfigAim {
  message Recoil {
    bool enabled = 1;
    int32 duration = 2;
    int32 factor = 3;
  }

  message Base {
    message Smooth {
      float min = 1;
      float max = 2;
    }

    bool enabled = 1;
    string key = 3;
    Smooth smooth_x = 4;
    Smooth smooth_y = 5;
  }

  int32 fov = 1;
  int32 offset_x = 2;
  int32 offset_y = 3;
  int32 input_delay = 4;
  Recoil recoil = 5;
  Base aim = 6;
  Base flick = 7;
}

enum RemoteConfigMode {
  Unknown = 0;
  PixelSeek = 1;
  ObjectDetection = 2;
}

message RemoteConfigCaptureMode {
  enum CaptureModeType {
    Unknown = 0;
    DXGI = 1;
    NDI = 2;
  }

  CaptureModeType type = 1;
  string source = 2; // For NDI
}

message RemoteConfigCapture {
  int32 size = 1;
  RemoteConfigCaptureMode mode = 2;
}

// RemoteConfig message
message RemoteConfig {
  RemoteConfigCapture capture = 1;
  RemoteConfigMode mode = 2;
  RemoteConfigAim aim = 3;
  RemoteConfigMouseServer mouse_server = 4;
}

message GetConfigRequest {
  string key = 1;
}

message GetConfigResponse {
  RemoteConfig remote_config = 1;
}

service License {
  rpc CreateSession(CreateSessionRequest) returns (CreateSessionResponse);
  rpc GetLatestSession(GetSessionRequest) returns (GetSessionResponse);
  rpc Ping(PingRequest) returns (PingResponse);
  rpc GetStatus(google.protobuf.Empty) returns (GetStatusResponse);
  rpc GetRemoteConfig(GetConfigRequest) returns (GetConfigResponse);
  rpc StreamConfig(GetConfigRequest) returns (stream GetConfigResponse);
}
